File: day7.py
import re

rules = {}
for line in open('day7.txt'):
    m = re.findall(r'(\d*)\s*(\w+ \w+) bags?', line)
    rules[m[0][1]] = {s: int(n) for n, s in m[1:] if s != 'no other'}

parents = {}
for src, bags in rules.items():
    for dst in bags.keys():
        parents.setdefault(dst, set()).add(src)

def all_parents(bag):
    res = parents.get(bag, set())
    for key in list(res): res.update(all_parents(key))
    return res
print(len(all_parents('shiny gold')))

def count_all(bag):
    return sum(v * count_all(k) for k, v in rules.get(bag, {}).items()) + 1
print(count_all('shiny gold') - 1)

File: day7.kt
import java.io.File

val rules = mutableMapOf<String,Map<String,Int>>()
val regex = Regex("""(\d*)\s*(\w+ \w+) bags?""")
File("aoc2020/day7.txt").forEachLine {
    val matches = regex.findAll(it).map { it.groupValues }.toList()
    rules[matches[0][2]] = matches.drop(1).mapNotNull {
        if (it[2] != "no other") it[2] to it[1].toInt() else null
    }.toMap()
}

val parents = mutableMapOf<String,MutableSet<String>>()
for ((src, bags) in rules) {
    for (dst in bags.keys) {
        parents.getOrPut(dst) { mutableSetOf<String>() }.add(src)
    }
}

fun all_parents(bag: String): Set<String> {
    if (!parents.contains(bag)) return emptySet<String>()
    var res = parents.getValue(bag).toList()
    return (res + res.flatMap { all_parents(it) }).toSet()
}
println(all_parents("shiny gold").size)

fun count_all(bag: String): Int {
    if (!rules.contains(bag)) return 1
    return rules.getValue(bag).map { it.value * count_all(it.key) }.sum() + 1
}
println(count_all("shiny gold") - 1)
